packages/catalog-client/src/CatalogClient.test.ts

```diff
@@ -142,12 +142,47 @@ describe('CatalogClient', () => {
       expect(response.items).toEqual([]);
     });
 
+    it('builds search filters property even those with URL unsafe values', async () => {
+      const mockedEndpoint = jest
+        .fn()
+        .mockImplementation((_req, res, ctx) =>
+          res(ctx.json({ items: [], totalItems: 0 })),
+        );
+
+      server.use(rest.get(`${mockBaseUrl}/entities/by-query`, mockedEndpoint));
+
+      const response = await client.queryEntities(
+        {
+          filter: [
+            {
+              '!@#$%': 't?i=1&a:2',
+              '^&*(){}[]': ['t%^url*encoded2', 'url'],
+            },
+          ],
+        },
+        { token },
+      );
+
+      expect(response).toEqual({ items: [], totalItems: 0 });
+      expect(mockedEndpoint).toHaveBeenCalledTimes(1);
+      // Validate that the URL is _actually_ encoded.
+      expect(mockedEndpoint.mock.calls[0][0].url.search).toBe(
+        '?filter=%21%40%23%24%25%3Dt%3Fi%3D1%26a%3A2%2C%5E%26%2A%28%29%7B%7D%5B%5D%3Dt%25%5Eurl%2Aencoded2%2C%5E%26%2A%28%29%7B%7D%5B%5D%3Durl',
+      );
+
+      // Validate that the URL matches the expected decoded value.
+      expect(
+        decodeURIComponent(mockedEndpoint.mock.calls[0][0].url.search),
+      ).toBe('?filter=!@#$%=t?i=1&a:2,^&*(){}[]=t%^url*encoded2,^&*(){}[]=url');
+    });
+
     it('builds entity field selectors properly', async () => {
-      expect.assertions(2);
+      expect.assertions(3);
 
       server.use(
         rest.get(`${mockBaseUrl}/entities`, (req, res, ctx) => {
-          expect(decodeURIComponent(req.url.search)).toBe('?fields=a.b,%C3%B6');
```

Comment:
When I originally adjusted this to use `decodeURIComponent`, it hid the underlying issue of the double encoding. The `%C3%B6` in this line. Adjusted this to be a two line test to ensure the encoding is happening and that the decoded value matches what we expect.

Reply:
I think it would be worthwhile to parse the query using the standard URLSearchParams, and then assert that THAT results in the expected output. Comparing the query string itself doesn't test it fully.

Reply:
https://github.com/backstage/backstage/pull/21719/commits/634f62065f290b663e310615657ac847f348b7e2

Github Copilot is eerily good at the replacement logic.